napi_js_props.md

```diff
@@ -0,0 +1,283 @@
+
+N-API exposes a set of APIs to interrogate and set properties on JavaScript objects.
+Some of these types are documented under [Section 7](https://tc39.github.io/ecma262/#sec-operations-on-objects) 
+of the [ECMAScript Language Specification](https://tc39.github.io/ecma262/).
+
+Fundamentally, all property keys in N-API can be represented in one of the following forms:
+- Named: a simple utf8 encoded string
+- Integer-Indexed: an index value represented by `uint32_t`
+- JavaScript Object: a JavaScript Object represented by `napi_value`. This can be a `napi_value` representing a String, Number or Symbol.
+
+N-API values are represented by the type `napi_value`.
+Any N-API call that requires a JavaScript object takes in a `napi_value`.
+However, it's the caller's responsibility to make sure that the 
+`napi_value` in question is of the JavaScript type expected by the API.
+
+## Structures
+
+### *napi_property_attributes*
+
+#### Definition
+```
+typedef enum {
+  napi_default = 0,
+  napi_read_only = 1 << 0,
+  napi_dont_enum = 1 << 1,
+  napi_dont_delete = 1 << 2,
+  napi_static_property = 1 << 10,
+} napi_property_attributes;
+```
+
+### Description
+`napi_property_attributes` are flags used to control the behavior of properties set on a JavaScript object. They roughly correspond to the attributes listed in [Section 6.1.7.1](https://tc39.github.io/ecma262/#table-2) of the [ECMAScript Language Specification](https://tc39.github.io/ecma262/).
+
+`napi_default` - Used to indicate that no explicit attributes are set on the given property
+`napi_read_only` - Used to indicate that a given property is not Writable
+`napi_dont_enum` - Used to indicate that a given property is not Enumerable
+`napi_dont_delete` - Used to indicate that a given property is not Configurable, as defined in [Section 6.1.7.1](https://tc39.github.io/ecma262/#table-2) of the [ECMAScript Language Specification](https://tc39.github.io/ecma262/).
+`napi_static_property` - Used to indicate that the property will be defined as a static property on a class as opposed to an instance property, which is the default. This is used only by `napi_define_class`. It is ignored by `napi_define_properties`
+
+### *napi_property_descriptor*
+
+#### Definition
+```
+typedef struct {
+  const char* utf8name;
+
+  napi_callback method;
+  napi_callback getter;
+  napi_callback setter;
+  napi_value value;
+
+  napi_property_attributes attributes;
+  void* data;
+} napi_property_descriptor;
+```
+
+#### Members
+- `utf8name`: String describing the key for the property, encoded as UTF8
+- `value`: The value that's retrieved by a get access of the property if the property is a data. If this is passed in, set `getter`, `setter`, `method` and `data` to `null` (since these members won't be used)
+- `getter`: A function to call when a get access of the property is performed. If this is passed in, set `value`, `method` and `data` to `null` (since these members won't be used)
```

Comment:
Maybe make it more specific that these get called when you call getters/settes from JavaScript as opposed to the N-API.  Same goes for the setter.